### PR TITLE
Promote configurable pod resolv.conf to Beta and add an e2e test

### LINKED ISSUE #56521: Need e2e test for configurable pod resolv.conf (a.k.a CustomPodDNS)
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug
/sig network

**What happened**:
From https://github.com/kubernetes/features/issues/504.

https://github.com/kubernetes/kubernetes/pull/55848 adds the Alpha support for Custom Pod DNS, given this is an opt-in feature that will not be triggered by any existing e2e test, we should write a explicit test to ensure it is and will be continue working.

/assign

### PR DESCRIPTION
**Pull request by MrHohn**
**What this PR does / why we need it**:
Feature issue: https://github.com/kubernetes/features/issues/504

There is no semantic changes. `CustomPodDNS` feature gate will be turned on by default.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #56521

**Special notes for your reviewer**:
/assign @bowei @thockin 

**Release note**:

```release-note
Adds BETA support for `DNSConfig` field in PodSpec and `DNSPolicy=None`.
```

### CODE REVIEW

**Comment by BenTheElder** [comment]
[remote caching experiment, please ignore]
/test pull-kubernetes-bazel-test-canary
/test pull-kubernetes-bazel-build-canary

File: pkg/kubelet/network/dns/dns_test.go [hunk 1]
```diff
@@ -247,10 +247,10 @@ func TestMergeDNSOptions(t *testing.T) {
 	}
 
 	for _, tc := range testCases {
-		options := mergeDNSOptions(tc.existingDNSConfigOptions, tc.dnsConfigOptions)
+		options := MergeDNSOptions(tc.existingDNSConfigOptions, tc.dnsConfigOptions)
```
Comment (bowei) [inline pkg/kubelet/network/dns/dns_test.go]: Can we do the test without referencing the functions that are used to implement the feature? This seems a bit circular dep (suppose MergeDNSOptions has a bug...).

Comment (MrHohn) [reply pkg/kubelet/network/dns/dns_test.go]: Indeed, will look into removing that dependency.

**Comment by MrHohn** [comment]
@bowei Rewrote the e2e test and moved most of previous resolv.conf checkings into unit test, PTAL thanks!

**Comment by thockin** [comment]
/approve

leaving the LGTM to Bowei.

File: test/e2e/network/dns.go [hunk 2]
```diff
@@ -444,11 +211,111 @@ var _ = SIGDescribe("DNS", func() {
...
-		pod3 := createDNSPod(f.Namespace.Name, wheezyProbeCmd, jessieProbeCmd)
+		pod3 := createDNSPod(f.Namespace.Name, wheezyProbeCmd, jessieProbeCmd, dnsTestPodHostName, dnsTestServiceName)
 
 		svc, err := f.ClientSet.CoreV1().Services(f.Namespace.Name).Get(externalNameService.Name, metav1.GetOptions{})
 		Expect(err).NotTo(HaveOccurred())
 
 		validateTargetedProbeOutput(f, pod3, []string{wheezyFileName, jessieFileName}, svc.Spec.ClusterIP)
 	})
+
+	It("should support configurable pod resolv.conf", func() {
```
Comment (MrHohn) [inline test/e2e/network/dns.go]: Most of the other parts are moving codes. Real change is in here.

**Pull request merged**